commit ef246a1480cc484cd2aeda75737cb0848616ddf3
Author: Johannes Berg <johannes.berg@intel.com>
Date:   Wed Sep 20 21:25:26 2023 +0300

    wifi: mac80211: support antenna control in injection
    
    Support antenna control for injection by parsing the antenna
    radiotap field (which may be presented multiple times) and
    telling the driver about the resulting antenna bitmap. Of
    course there's no guarantee the driver will actually honour
    this, just like any other injection control.
    
    If misconfigured, i.e. the injected HT/VHT MCS needs more
    chains than antennas are configured, the bitmap is reset to
    zero, indicating no selection.
    
    For now this is only set up for two anntenas so we keep more
    free bits, but that can be trivially extended if any driver
    implements support for it that can deal with hardware with
    more antennas.
    
    Signed-off-by: Johannes Berg <johannes.berg@intel.com>
    Signed-off-by: Gregory Greenman <gregory.greenman@intel.com>
    Link: https://lore.kernel.org/r/20230920211508.f71001aa4da9.I00ccb762a806ea62bc3d728fa3a0d29f4f285eeb@changeid
    Signed-off-by: Johannes Berg <johannes.berg@intel.com>

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index d4ef2a605cb4..72375eceb786 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -1178,7 +1178,11 @@ struct ieee80211_tx_info {
 					u8 use_cts_prot:1;
 					u8 short_preamble:1;
 					u8 skip_table:1;
-					/* 2 bytes free */
+
+					/* for injection only (bitmap) */
+					u8 antennas:2;
+
+					/* 14 bits free */
 				};
 				/* only needed before rate control */
 				unsigned long jiffies;